File: 14_stacks_and_queues_2/06_min_stack.py
class MinStack:
    def __init__(self):
        self.stack = []
        self.mini_stack = []

    def push(self, val: int) -> None:
        self.stack.append(val)
        if self.mini_stack:
            self.mini_stack.append(min(self.mini_stack[-1], val))
        else:
            self.mini_stack.append(val)

    def pop(self) -> None:
        self.stack.pop()
        self.mini_stack.pop()

    def top(self) -> int:
        return self.stack[-1]

    def getMin(self) -> int:
        return self.mini_stack[-1]

File: 14_stacks_and_queues_2/06_min_stack.rs
use std::cmp::min;

struct MinStack {
    stack: Vec<i32>,
    min_stack: Vec<i32>,
}

impl MinStack {
    fn new() -> Self {
        MinStack {
            stack: Vec::new(),
            min_stack: Vec::new(),
        }
    }

    fn push(&mut self, val: i32) {
        if self.stack.is_empty() {
            self.min_stack.push(val);
        } else {
            let top = self.min_stack.last().unwrap();
            self.min_stack.push(min(*top, val));
        }
        self.stack.push(val);
    }

    fn pop(&mut self) {
        self.stack.pop();
        self.min_stack.pop();
    }

    fn top(&self) -> i32 {
        *self.stack.last().unwrap()
    }

    fn get_min(&self) -> i32 {
        *self.min_stack.last().unwrap()
    }
}
